streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1677,6 +1677,17 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
       } yield pull
     }
 
+  /**
+   * Halts the evaluation of this stream when the provided IO completes. The given IO
+   * will be evaluated as part of this stream, and its success will be discarded.
+   *
+   * If the IO completes with a failure, the stream will emit that failure.
+   */
+  final def haltWhen[E1 >: E](io: IO[E1, Unit]): ZStream[R, E1, A] =
```

Comment:
There is probably a better way to do this one

Reply:
Hmm. Why don't we implement `haltWhen(Promise)` in terms of this combinator? I think it should be simpler.

Reply:
Agreed, just didn't want to disrupt the existing logic (which depends on `poll`) too much.

Reply:
Oh, I forgot it does that. In that case your implementation may be the simpler one. Let me know if so and happy to merge as is.

Reply:
I *think* an alternate implementation of `haltWhen(IO)` could be like this, and then `haltWhen(Promise)` could be implemented in terms of `haltWhen(p.await)`. But I'm probably missing some subtlety.

```scala
  final def haltWhen[E1 >: E](io: IO[E1, Unit]): ZStream[R, E1, A] =
    ZStream {
      for {
        as    <- self.process
        runIO <- (io.asSomeError *> Pull.end).forkManaged
      } yield as.raceFirst(runIO.join)
    }
```

Reply:
Actually, this implementation fails the test because it doesn't explicitly check the IO *first*, which the test expects because it is expecting an already-successful `ZStream` to fail if the `IO` has failed. Not sure if that's more an issue with the expectation or the implementation (or maybe that's the subtlety I'm missing 😁 )

Reply:
Good catch! The implementation in this revision looks good. Let’s just generalize it to `[R1 <: R, E1 >: E](zio: ZIO[R1, E1, Any])`.